tools/autograd/gen_python_functions.py

```diff
@@ -58,6 +58,31 @@
   ${unpack_self}
   ParsedArgs<${max_args}> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
+
+  ${declare_namedtuple_return_types}
+  ${dispatch}
+
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+""")
+
+# Check for __torch_function__ overrides
+PY_VARIABLE_FUNCTION_VARARGS = CodeTemplate("""\
+static PyObject * ${pycname}(PyObject* self_, PyObject* args, PyObject* kwargs)
+{
+  HANDLE_TH_ERRORS
+  static PythonArgParser parser({
+    ${signatures}
+  }, /*traceable=*/${traceable});
+  ${unpack_self}
+  ParsedArgs<${max_args}> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+
+  if(r.has_torch_function()){
+    PyObject* torch_function = PyObject_FastGetAttrString(r.get_overloaded_arg(0), "__torch_function__");
+    return PyObject_CallFunctionObjArgs(torch_function, PyUnicode_FromString(r.get_func_name().data()), args, kwargs, NULL);
```

Comment:
I'm curious: I assume passing a string function name here is because we're copying the Numpy design, but I wonder, why are we passing a string, rather than, e.g., the actual function that was invoked (e.g., do a `func == torch.add` test)

Reply:
NumPy actually passes the function, I think we should do that as well.

See e.g. https://docs.scipy.org/doc/numpy/reference/arrays.classes.html#numpy.class.__array_function__